更新主题出现 Duplicate entry '/registry/settings/theme-earth-setting' 的问题

### 是什么版本出现了此问题？

2.0.1-SNAPSHOT

### 使用的什么数据库？

H2

### 使用的哪种方式部署？

Docker

### 在线站点地址

_No response_

### 发生了什么？

删除数据库，重新启动，在更新主题时发现这个问题

### 相关日志输出

```shell

org.springframework.r2dbc.UncategorizedR2dbcException: executeMany; SQL [INSERT INTO extensions (name, data, version) VALUES (?, ?, ?)]; Duplicate entry '/registry/settings/theme-earth-setting' for key 'PRIMARY'
	at org.springframework.r2dbc.connection.ConnectionFactoryUtils.convertR2dbcException(ConnectionFactoryUtils.java:238) ~[spring-r2dbc-6.0.2.jar:6.0.2]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ org.springframework.web.filter.reactive.ServerHttpObservationFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.console.ConsoleProxyFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthorizationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ExceptionTranslationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ LogoutWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ServerRequestCacheWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityContextServerWebExchangeWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ LogoutPageGeneratingWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ LoginPageGeneratingWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AnonymousAuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ReactorContextWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ CsrfWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ CorsWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ HttpHeaderWriterWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ServerWebExchangeReactorContextWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ org.springframework.security.web.server.WebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ HTTP POST "/apis/api.console.halo.run/v1alpha1/themes/theme-earth/upgrade" [ExceptionHandlingWebHandler]
Original Stack Trace:
		at org.springframework.r2dbc.connection.ConnectionFactoryUtils.convertR2dbcException(ConnectionFactoryUtils.java:238) ~[spring-r2dbc-6.0.2.jar:6.0.2]
		at org.springframework.r2dbc.core.DefaultDatabaseClient.lambda$inConnectionMany$8(DefaultDatabaseClient.java:150) ~[spring-r2dbc-6.0.2.jar:6.0.2]
		at reactor.core.publisher.Flux.lambda$onErrorMap$28(Flux.java:7123) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.Flux.lambda$onErrorResume$29(Flux.java:7176) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxUsingWhen$UsingWhenSubscriber.deferredError(FluxUsingWhen.java:398) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxUsingWhen$RollbackInner.onComplete(FluxUsingWhen.java:475) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:2168) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:2168) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onComplete(FluxDoFinally.java:128) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onComplete(FluxDoFinally.java:128) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onComplete(MonoIgnoreElements.java:89) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:121) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:64) ~[reactor-core-3.5.0.jar:3.5.0]
		at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147) ~[na:na]
		at com.github.jasync.sql.db.mysql.MySQLConnection$close$lambda-2$$inlined$onCompleteAsync$1.accept(FutureUtils.kt:110) ~[jasync-mysql-2.1.7.jar:na]
		at com.github.jasync.sql.db.mysql.MySQLConnection$close$lambda-2$$inlined$onCompleteAsync$1.accept(FutureUtils.kt:92) ~[jasync-mysql-2.1.7.jar:na]
		at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483) ~[na:na]
		at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:373) ~[na:na]
		at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java) ~[na:na]
		at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182) ~[na:na]
		at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) ~[na:na]
		at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622) ~[na:na]
		at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165) ~[na:na]
Caused by: com.github.jasync.r2dbc.mysql.JasyncDatabaseException: Duplicate entry '/registry/settings/theme-earth-setting' for key 'PRIMARY'
	at com.github.jasync.r2dbc.mysql.JasyncStatement$execute$3.invoke(JasyncStatement.kt:129) ~[jasync-r2dbc-mysql-2.1.7.jar:na]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ SQL "INSERT INTO extensions (name, data, version) VALUES (?, ?, ?)" [DatabaseClient]
Original Stack Trace:
		at com.github.jasync.r2dbc.mysql.JasyncStatement$execute$3.invoke(JasyncStatement.kt:129) ~[jasync-r2dbc-mysql-2.1.7.jar:na]
		at com.github.jasync.r2dbc.mysql.JasyncStatement$execute$3.invoke(JasyncStatement.kt:100) ~[jasync-r2dbc-mysql-2.1.7.jar:na]
		at reactor.kotlin.core.publisher.FluxExtensionsKt.onErrorMap$lambda-1(FluxExtensions.kt:171) ~[reactor-kotlin-extensions-1.2.0.jar:1.2.0]
		at reactor.core.publisher.Flux.lambda$onErrorMap$28(Flux.java:7123) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.Flux.lambda$onErrorResume$29(Flux.java:7176) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:134) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.maybeOnError(FluxConcatMapNoPrefetch.java:326) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxConcatMapNoPrefetch$FluxConcatMapNoPrefetchSubscriber.innerError(FluxConcatMapNoPrefetch.java:297) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.FluxConcatMap$ConcatMapInner.onError(FluxConcatMap.java:875) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:117) ~[reactor-core-3.5.0.jar:3.5.0]
		at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:64) ~[reactor-core-3.5.0.jar:3.5.0]
		at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[na:na]
		at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[na:na]
		at com.github.jasync.sql.db.util.FutureUtilsKt.failed(FutureUtils.kt:18) ~[jasync-pool-2.1.7.jar:na]
		at com.github.jasync.sql.db.mysql.MySQLConnection.failQueryPromise$lambda-11(MySQLConnection.kt:355) ~[jasync-mysql-2.1.7.jar:na]
		at java.base/java.util.Optional.ifPresent(Optional.java:178) ~[na:na]
		at com.github.jasync.sql.db.mysql.MySQLConnection.failQueryPromise(MySQLConnection.kt:354) ~[jasync-mysql-2.1.7.jar:na]
		at com.github.jasync.sql.db.mysql.MySQLConnection.setException(MySQLConnection.kt:205) ~[jasync-mysql-2.1.7.jar:na]
		at com.github.jasync.sql.db.mysql.MySQLConnection.onError(MySQLConnection.kt:199) ~[jasync-mysql-2.1.7.jar:na]
		at com.github.jasync.sql.db.mysql.codec.MySQLConnectionHandler.channelRead0(MySQLConnectionHandler.kt:125) ~[jasync-mysql-2.1.7.jar:na]
		at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.85.Final.jar:4.1.85.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.85.Final.jar:4.1.85.Final]
		at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: com.github.jasync.sql.db.mysql.exceptions.MySQLException: Error 1062 - #23000 - Duplicate entry '/registry/settings/theme-earth-setting' for key 'PRIMARY'
	at com.github.jasync.sql.db.mysql.MySQLConnection.onError(MySQLConnection.kt:198) ~[jasync-mysql-2.1.7.jar:na]
	at com.github.jasync.sql.db.mysql.codec.MySQLConnectionHandler.channelRead0(MySQLConnectionHandler.kt:125) ~[jasync-mysql-2.1.7.jar:na]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.85.Final.jar:4.1.85.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.85.Final.jar:4.1.85.Final]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
```


### 附加信息

_No response_

--- a/src/main/java/run/halo/app/core/extension/reconciler/ThemeReconciler.java
+++ b/src/main/java/run/halo/app/core/extension/reconciler/ThemeReconciler.java
@@ -8,6 +8,7 @@ import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
 import org.apache.commons.lang3.StringUtils;
+import org.springframework.retry.support.RetryTemplate;
 import org.springframework.stereotype.Component;
 import org.springframework.util.CollectionUtils;
 import org.springframework.util.FileSystemUtils;
@@ -40,6 +41,12 @@ public class ThemeReconciler implements Reconciler<Request> {
 
     private final ExtensionClient client;
     private final ThemePathPolicy themePathPolicy;
+
+    private final RetryTemplate retryTemplate = RetryTemplate.builder()
+        .maxAttempts(20)
+        .fixedBackoff(300)
+        .retryOn(IllegalStateException.class)
+        .build();
 
     public ThemeReconciler(ExtensionClient client, HaloProperties haloProperties) {
         this.client = client;
@@ -157,20 +164,39 @@ public class ThemeReconciler implements Reconciler<Request> {
         if (StringUtils.isNotBlank(settingName)) {
             client.fetch(Setting.class, settingName)
                 .ifPresent(client::delete);
+            retryTemplate.execute(callback -> {
+                client.fetch(Setting.class, settingName).ifPresent(setting -> {
+                    throw new IllegalStateException("Waiting for setting to be deleted.");
+                });
+                return null;
+            });
         }
         // delete annotation setting
         deleteAnnotationSettings(theme.getMetadata().getName());
     }
 
     private void deleteAnnotationSettings(String themeName) {
-        List<AnnotationSetting> result = client.list(AnnotationSetting.class, annotationSetting -> {
-            Map<String, String> labels = ExtensionUtil.nullSafeLabels(annotationSetting);
-            return themeName.equals(labels.get(Theme.THEME_NAME_LABEL));
-        }, null);
+        List<AnnotationSetting> result = listAnnotationSettingsByThemeName(themeName);
 
         for (AnnotationSetting annotationSetting : result) {
             client.delete(annotationSetting);
         }
+
+        retryTemplate.execute(callback -> {
+            List<AnnotationSetting> annotationSettings =
+                listAnnotationSettingsByThemeName(themeName);
+            if (annotationSettings.isEmpty()) {
+                return null;
+            }
+            throw new IllegalStateException("Waiting for annotation settings to be deleted.");
+        });
+    }
+
+    private List<AnnotationSetting> listAnnotationSettingsByThemeName(String themeName) {
+        return client.list(AnnotationSetting.class, annotationSetting -> {
+            Map<String, String> labels = ExtensionUtil.nullSafeLabels(annotationSetting);
+            return themeName.equals(labels.get(Theme.THEME_NAME_LABEL));
+        }, null);
     }
 
     private void deleteThemeFiles(Theme theme) {

--- a/src/test/java/run/halo/app/core/extension/reconciler/ThemeReconcilerTest.java
+++ b/src/test/java/run/halo/app/core/extension/reconciler/ThemeReconcilerTest.java
@@ -1,6 +1,7 @@
 package run.halo.app.core.extension.reconciler;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.times;
@@ -12,6 +13,7 @@ import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.time.Instant;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import org.json.JSONException;
@@ -23,7 +25,9 @@ import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.mockito.Mockito;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.stubbing.Answer;
 import org.skyscreamer.jsonassert.JSONAssert;
+import org.springframework.retry.RetryException;
 import org.springframework.util.FileSystemUtils;
 import org.springframework.util.ResourceUtils;
 import run.halo.app.core.extension.AnnotationSetting;
@@ -32,6 +36,7 @@ import run.halo.app.core.extension.Theme;
 import run.halo.app.extension.ConfigMap;
 import run.halo.app.extension.ExtensionClient;
 import run.halo.app.extension.Metadata;
+import run.halo.app.extension.MetadataOperator;
 import run.halo.app.extension.controller.Reconciler;
 import run.halo.app.infra.properties.HaloProperties;
 import run.halo.app.infra.utils.JsonUtils;
@@ -105,12 +110,98 @@ class ThemeReconcilerTest {
         themeReconciler.reconcile(new Reconciler.Request(metadata.getName()));
 
         verify(extensionClient, times(2)).fetch(eq(Theme.class), eq(metadata.getName()));
-        verify(extensionClient, times(1)).fetch(eq(Setting.class), eq(themeSpec.getSettingName()));
+        verify(extensionClient, times(2)).fetch(eq(Setting.class), eq(themeSpec.getSettingName()));
 
-        verify(extensionClient, times(1)).list(eq(AnnotationSetting.class), any(), any());
+        verify(extensionClient, times(2)).list(eq(AnnotationSetting.class), any(), any());
 
         assertThat(Files.exists(testWorkDir)).isTrue();
         assertThat(Files.exists(defaultThemePath)).isFalse();
+    }
+
+    @Test
+    void reconcileDeleteRetry() {
+        Theme theme = fakeTheme();
+        final MetadataOperator metadata = theme.getMetadata();
+
+        Path testWorkDir = tempDirectory.resolve("reconcile-delete");
+        when(haloProperties.getWorkDir()).thenReturn(testWorkDir);
+
+        final ThemeReconciler themeReconciler =
+            new ThemeReconciler(extensionClient, haloProperties);
+
+        final int[] retryFlags = {0, 0};
+        when(extensionClient.fetch(eq(Setting.class), eq("theme-test-setting")))
+            .thenAnswer((Answer<Optional<Setting>>) invocation -> {
+                retryFlags[0]++;
+                // retry 2 times
+                if (retryFlags[0] < 3) {
+                    return Optional.of(new Setting());
+                }
+                return Optional.empty();
+            });
+
+        when(extensionClient.list(eq(AnnotationSetting.class), any(), eq(null)))
+            .thenAnswer((Answer<List<AnnotationSetting>>) invocation -> {
+                retryFlags[1]++;
+                // retry 2 times
+                if (retryFlags[1] < 3) {
+                    return List.of(new AnnotationSetting());
+                }
+                return List.of();
+            });
+
+        themeReconciler.reconcile(new Reconciler.Request(metadata.getName()));
+
+        String settingName = theme.getSpec().getSettingName();
+        verify(extensionClient, times(2)).fetch(eq(Theme.class), eq(metadata.getName()));
+        verify(extensionClient, times(3)).fetch(eq(Setting.class), eq(settingName));
+        verify(extensionClient, times(3)).list(eq(AnnotationSetting.class), any(), eq(null));
+    }
+
+    @Test
+    void reconcileDeleteRetryWhenThrowException() {
+        Theme theme = fakeTheme();
+
+        Path testWorkDir = tempDirectory.resolve("reconcile-delete");
+        when(haloProperties.getWorkDir()).thenReturn(testWorkDir);
+
+        final ThemeReconciler themeReconciler =
+            new ThemeReconciler(extensionClient, haloProperties);
+
+        final int[] retryFlags = {0};
+        when(extensionClient.fetch(eq(Setting.class), eq("theme-test-setting")))
+            .thenAnswer((Answer<Optional<Setting>>) invocation -> {
+                retryFlags[0]++;
+                // retry 2 times
+                if (retryFlags[0] < 2) {
+                    return Optional.of(new Setting());
+                }
+                throw new RetryException("retry exception.");
+            });
+
+        String settingName = theme.getSpec().getSettingName();
+        assertThatThrownBy(
+            () -> themeReconciler.reconcile(new Reconciler.Request(theme.getMetadata().getName())))
+            .isInstanceOf(RetryException.class)
+            .hasMessage("retry exception.");
+
+        verify(extensionClient, times(2)).fetch(eq(Setting.class), eq(settingName));
+    }
+
+    private Theme fakeTheme() {
+        Theme theme = new Theme();
+        Metadata metadata = new Metadata();
+        metadata.setName("theme-test");
+        metadata.setDeletionTimestamp(Instant.now());
+        theme.setMetadata(metadata);
+        theme.setKind(Theme.KIND);
+        theme.setApiVersion("theme.halo.run/v1alpha1");
+        Theme.ThemeSpec themeSpec = new Theme.ThemeSpec();
+        themeSpec.setSettingName("theme-test-setting");
+        theme.setSpec(themeSpec);
+        when(extensionClient.fetch(eq(Theme.class), eq(metadata.getName())))
+            .thenReturn(Optional.of(theme));
+        return theme;
     }
 
     @Test
@@ -169,10 +260,10 @@ class ThemeReconcilerTest {
         ConfigMap defaultValueConfigMap = configMapCaptor.getValue();
         Map<String, String> data = defaultValueConfigMap.getData();
         JSONAssert.assertEquals("""
-            {
-                "sns": "{\\"email\\":\\"example@exmple.com\\"}"
-            }
-            """,
+                {
+                    "sns": "{\\"email\\":\\"example@exmple.com\\"}"
+                }
+                """,
             JsonUtils.objectToJson(data),
             true);
     }

